Add Session 4 and 10 exercises. (#9)
Lumi-supercomputer/Getting_Started_with_AI_workshop: 04_Understanding_GPU_activity_and_checking_jobs/README.md

# 04 Understanding GPU activity & checking jobs

These examples are based on the Pytorch container provided to you to do training at:
```
/appl/local/containers/sif-images/lumi-pytorch-rocm-5.6.1-python-3.10-pytorch-v2.2.2.sif 
```
This container was extended with some components required by the examples, notoriously the HuggingFace transformers module, resulting in the container in:
```

> /scratch/project_465001063/containers/pytorch_transformers.sif

Reviewer: This container is actually built using cotainr from the [pytorch_transformers.yml](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/blob/main/bonus_material/exercise_container_recipes/pytorch_transformers.yml) conda environment and is based on the /appl/local/containers/sif-images/lumi-rocm-rocm-5.6.1.sif base image.

Author: Ok, corrected.

```
This is very similiar to what you have been doing with `sbatch` should you be using a run script with:
```
#SBATCH --account=project_465001063
#SBATCH --reservation=AI_workshop
#SBATCH --partition=small-g
#SBATCH --gpus-per-node=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=7
#SBATCH --mem-per-gpu=60G
#SBATCH --time=0:30:00
```
The difference is that it gives you a mechanism to just allocate the nodes without running anything. You can then issue `srun` commands interactively which can be useful to experiment more easily. You are always welcome to transition to use `sbatch` if that is preferred.

We'll also leverage the configuration for singularity provided by:
```
module purge
module use /appl/local/training/modules/AI-20240529/
module load singularity-userfilesystems singularity-CPEbits
``` 

With the allocation and container set we can do a quick smoke test to make sure Pytorch can detect the GPUs available in a node:
```
srun singularity exec \
    /scratch/project_465001063/containers/pytorch_transformers.sif \
    bash -c '$WITH_CONDA ; \
             python -c "import torch; print(torch.cuda.device_count())"'
```

> It should yield `1` given that only one GPU was requested. Note that each time a node is used for the first time, there is a latency to have the cointainer loaded. Running the command above again on the same allocation should complete faster.

Reviewer: ```suggestion
It should yield `1` given that only one GPU was requested. Note that each time a node is used for the first time, there is a latency to have the container loaded. Running the command above again on the same allocation should complete faster.
```

Author: Fixed.

## Hands-on exercise

We will leverage here the same LLM example as before with small adaptations. No extra files are needed. You might be interested in collating the different steps in a batch script or run interactively as presented. 

### 1. Let's recover our LLM example.
Here we'll recover our fine-tunning example for IMDB movie review generation:

```

> curl -o GPT-neo-IMDB-finetuning.py -L https://raw.githubusercontent.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/main/03_Your_first_AI_training_job_on_LUMI/reference_solution/GPT-neo-IMDB-finetuning.py

Reviewer: In the other lectures, we have just included a copy of the example script in the lecture folder.

### 3. Monitoring GPU activity

Monitoring in a separate tab can be done by checking you jobID and connect to the first node of the allocation. E.g.:

* Get jobID - in this case `7100665`:
```
squeue --me
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
           7100665   small-g interact samantao  R    1:03:21      1 nid005021
           ...
```
* Start interactive parallel session:
```
srun --jobid 7100665 --interactive --pty /bin/bash
```
* Use `rocm-smi` to monitor GPU activity:
```
watch -n1 rocm-smi
```
This will give a snapshot of the GPU utilization captured by the driver every second:
```
======================= ROCm System Management Interface =======================
================================= Concise Info =================================
GPU  Temp   AvgPwr  SCLK     MCLK     Fan  Perf    PwrCap  VRAM%  GPU%  
0    58.0c  324.0W  1650Mhz  1600Mhz  0%   manual  500.0W   98%   100%  
1    49.0c  N/A     800Mhz   1600Mhz  0%   manual  0.0W      0%   0%    
================================================================================
============================= End of ROCm SMI Log ==============================
```

> As expected we only have activity on one GCD but the power metrics are per GPU. Note that these numbers needs to be interpreted. For example, if `GPU%` shows `100%` that doesn NOT necessarily mean the GPU is being well utilized. A better metric is drawn power `AvgPwr`: oscilating around `500.0W` is an indication there is significant compute activity on the full GPU.

Reviewer: ```suggestion
As expected we only have activity on one GCD but the power metrics are per GPU. Note that these numbers needs to be interpreted. For example, if `GPU%` shows `100%` that does NOT necessarily mean the GPU is being well utilized. A better metric is drawn power `AvgPwr`: oscillating around `500.0W` is an indication there is significant compute activity on the full GPU.
```

Author: Fixed.

> Here we see drawn power to oscilate around `300.0W` while a single GCD is being used, which is an indication that we might be compute bound.

Reviewer: ```suggestion
Here we see drawn power to oscillate around `300.0W` while a single GCD is being used, which is an indication that we might be compute bound.
```

Author: Fixed
